### PR TITLE
Split KeyPair from Node to Validator Config

### PR DESCRIPTION
**Pull request by kchulj**
Currently, both full nodes and validators share a common `NodeConfig`, and are distinguished by the `is_validator` boolean.
However, a full node should not have a `KeyPair` (and other validator-related attributes) as its attribute.

As discussed, there are two ways of handling this issue:
1. Make a separate `ValidatorConfig` struct which contains attributes exclusive to a `Validator` **(current approach)**
2. Embed `NodeConfig` into a new `ValidatorConfig`

The current approach is ok, but gets messy in `Base.d` because `makeMainConfig` takes an extra `ValidatorConfig` as its parameter. The `ValidatorConfig` is also created for full nodes, just with initial values. The function could be overloaded to remove this parameter for full nodes.

In the 2nd approach, the main `Config` has to be templated to either take `NodeConfig` or `ValidatorConfig`. The disadvantage would be that more code changes will be required in agora as a whole.

Relates #816

### CODE REVIEW

**Comment by kchulj** [comment]
Ready for review.

File: doc/config.example.yaml [hunk 1]
```diff
@@ -37,15 +31,27 @@ node:
...
+  enabled: true
   # This is a randomly generated keypair
   # If this node is not a validator, this will be ignored
   #
   # DO NOT USE THOSE VALUES ANYWHERE
   # Private seed:    SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4
   # Public address:  GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ
   seed:    SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4
-  # Path to the data directory (if the path doesn't exist it will be created)
-  data_dir: .cache
```
Comment (AndrejMitrovic) [inline doc/config.example.yaml:48]: Unrelated change? The FullNode also needs a cache directory.

File: doc/config.example.yaml [hunk 2]
```diff
@@ -37,15 +31,27 @@ node:
...
+################################################################################
+##                             Validator configuration                        ##
+## The server can operate in two modes: full node and validator node.         ##
+## The full node does not participate in consensus,                           ##
+## it only replicates the state of the blockchain.                            ##
+################################################################################
+validator:
+  # Whether or not we should act as a validator
+  # When validating, the `seed` of an eligible account is required
+  # An eligible account has at least 4 units frozen in it
```
Comment (AndrejMitrovic) [inline doc/config.example.yaml]: It's 40k coins, not 4 units.

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -1355,21 +1355,19 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
...
-                validator_idx++, node_key.address.prettify);
-        else
-            address = format("FullNode #%s", full_node_idx++);
+    string validatorAddress (KeyPair node_key) {
+        return format("Validator #%s (%s)",
+            validator_idx++, node_key.address.prettify);
+    }
+    string fullNodeAddress () {
+        return format("FullNode #%s", full_node_idx++);
+    }
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]:
```suggestion
    string validatorAddress (KeyPair node_key) 
    {
        return format("Validator #%s (%s)",
            validator_idx++, node_key.address.prettify);
    }
    string fullNodeAddress () 
    {
        return format("FullNode #%s", full_node_idx++);
    }
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

File: doc/config.example.yaml (hunk 1 above)
Comment (kchulj) [reply doc/config.example.yaml:48]: For some reason, the `git diff` deleted it here and moved it to line 34-35.

Comment (AndrejMitrovic) [reply doc/config.example.yaml:48]: Oh I see!

**Review by AndrejMitrovic — APPROVED** [review]

File: source/agora/common/Config.d [hunk 4]
```diff
@@ -160,6 +157,16 @@ public struct NodeConfig
     public Duration preimage_reveal_interval = 10.seconds;
 }
 
+// Validator config
```
Comment (Geod24) [inline source/agora/common/Config.d]:
```suggestion
/// Validator config
```

To make sure it shows up in documentation ;)

File: source/agora/common/Config.d [hunk 5]
```diff
@@ -399,18 +394,52 @@ node:
...
     {
     immutable conf_example = `
-node:
-  is_validator: true
+validator:
+  enabled: true
 `;
         auto node = Loader.fromString(conf_example).load();
-        assertThrown!Exception(parseNodeConfig("node" in node, cmdln));
+        assertThrown!Exception(parseConfigImpl(cmdln, node));
```
Comment (Geod24) [inline source/agora/common/Config.d]:
```suggestion
        assertThrown!Exception(parseValidatorConfig("validator" in node, cmdln));
```

File: source/agora/node/Runner.d [hunk 6]
```diff
@@ -64,7 +64,7 @@ public NodeListenerTuple runNode (Config config)
     mkdirRecurse(config.node.data_dir);
 
     FullNode node;
-    if (config.node.is_validator)
+    if (config.validator.enabled == true)
```
Comment (Geod24) [inline source/agora/node/Runner.d]:
```suggestion
    if (config.validator.enabled)
```

File: source/agora/node/Validator.d [hunk 7]
```diff
@@ -80,13 +80,13 @@ public class Validator : FullNode, API
     /// Ctor
     public this (const Config config)
     {
-        assert(config.node.is_validator);
+        assert(config.validator.enabled == true);
```
Comment (Geod24) [inline source/agora/node/Validator.d]:
```suggestion
        assert(config.validator.enabled);
```

File: source/agora/test/Base.d [hunk 8]
```diff
@@ -639,7 +639,7 @@ public class TestAPIManager
     {
         RemoteAPI!TestAPI api;
         auto time = new shared(time_t)(this.initial_time);
-        if (conf.node.is_validator)
+        if (conf.validator.enabled == true)
```
Comment (Geod24) [inline source/agora/test/Base.d]:
```suggestion
        if (conf.validator.enabled)
```

File: source/agora/test/EnrollmentManager.d [hunk 9]
```diff
@@ -242,7 +242,7 @@ unittest
             if (this.nodes.length == 0)
             {
                 auto time = new shared(time_t)(this.initial_time);
-                assert(conf.node.is_validator);
+                assert(conf.validator.enabled == true);
```
Comment (Geod24) [inline source/agora/test/EnrollmentManager.d]:
```suggestion
                assert(conf.validator.enabled);
```

File: source/agora/test/NetworkManager.d [hunk 10]
```diff
@@ -138,7 +138,7 @@ unittest
             // 1 byzantine FullNode => lies about the blockchain
             //   (returns syntactically invalid data)
             // 1 good FullNode => it accepts only the valid blockchain
-            if (conf.node.is_validator)
+            if (conf.validator.enabled == true)
```
Comment (Geod24) [inline source/agora/test/NetworkManager.d]:
```suggestion
            if (conf.validator.enabled)
```

**Review by Geod24 — APPROVED** [review]
LGTM, just some nits

**Comment by Geod24** [comment]
Hum, the commit title / messages seems to have been lost in the rebase, I'll just use `squash` merge.

**Review by linked0 — APPROVED** [review]
LGTM
